apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -449,7 +449,10 @@ function checkData(data, origData, phase) {
     }
     if (!editPhases.includes(phase)) {
       if (!_.has(origData, prop)) return '"' + prop + '" is missing from "origData"';
-      if (!_.isEqual(data[prop], origData[prop])) {
+
+      // If `data` and `origData` are the same object, we can skip the expensive
+      // deep comparison of them. We rely on this optimization in `processQuestionHtml`.
+      if (data !== origData && !_.isEqual(data[prop], origData[prop])) {
```

Comment:
Would this capture other cases as well if there is a shallow copy at some point? I'm genuinely asking, it might not be beneficial at all.
```suggestion
      if (data[prop] !== origData[prop] && !_.isEqual(data[prop], origData[prop])) {
```

Reply:
This makes marginally more sense and would handle more hypothetical cases, though I don't think it changes anything given our current usages of this function. I'll accept this, thanks!